### PR TITLE
gate: add metric for wait

### LINKED ISSUE #11365: Gate needs a waiting duration metric
## Proposal

https://github.com/prometheus/prometheus/blob/main/util/gate/gate.go this is used in the remote read handler to limit number of concurrent requests. However, if we are close to that limit then currently it is impossible to know that. Ideally, we'd know if requests are being held up at the gate. I suggest adding a "waiting duration" histogram metric there so that users could understand whether they are hitting that limit.

### PR DESCRIPTION
**Pull request by GiedriusS**
Add a metric that shows how long we need to wait before a request in the remote read handler gets executed. It is useful to have it to understand whether the concurrency limit needs to be lifted or not.

Closes #11365.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

### CODE REVIEW

File: util/gate/gate.go [hunk 1]
```diff
@@ -13,23 +13,42 @@
...
 // being concurrently executed.
-func New(length int) *Gate {
+func New(length int, name string, r prometheus.Registerer) *Gate {
 	return &Gate{
 		ch: make(chan struct{}, length),
+		waitDuration: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Namespace:   "prometheus",
+			Subsystem:   "api",
+			Help:        "How long was the wait at the gate before the query execution",
+			Buckets:     prometheus.ExponentialBuckets(0.1, 2, 10),
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply util/gate/gate.go]: @juliusv thanks for the review and sorry for the delay. Your suggestion makes sense. I have updated my code.

**Comment by GiedriusS** [comment]
Very valid points! I have done B as it permits code reuse in projects and has maximum flexibility.

File: storage/remote/read_handler.go [hunk 2]
```diff
@@ -45,11 +46,17 @@ type readHandler struct {
...
-		remoteReadGate:            gate.New(remoteReadConcurrencyLimit),
+		logger:                logger,
+		queryable:             queryable,
+		config:                config,
+		remoteReadSampleLimit: remoteReadSampleLimit,
+		remoteReadGate: gate.New(remoteReadConcurrencyLimit, promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace:   "prometheus",
+			Subsystem:   "api",
+			Help:        "Total sum of how long was the wait at the gate before query execution",
+			Name:        "gate_total_wait_duration_seconds",
```
Comment (juliusv) [inline storage/remote/read_handler.go:57]: The structure of the code looks good now, but given that we are passing the metric into the gate, we can now also change the metric name and its description to be more specific to how the gate's usage. We have some precedent at https://github.com/prometheus/prometheus/blob/f55ab2217984770aa1eecd0f2d5f54580029b1c0/storage/remote/read_handler.go#L56-L61 for a remote-read-related metric already, so how about `prometheus_api_remote_read_wait_duration_seconds_total` (for the fully expanded metric name), with a description of `Total number of seconds spent waiting at the query gate before remote read query execution`?

**Comment by roidelapluie** [comment]
cc @GiedriusS Can you please address Julius' comment and rebase this pull request please? Thanks!

**Comment by beorn7** [comment]
Looking at this during the bug scrub again. 

We'll close this for now as there was no follow-up, but we would love to see @GiedriusS (or anyone else) to pick up the work.

**Pull request closed without merging**